### PR TITLE
Loss vs Cost graph

### PR DESCRIPTION
**Pull request by andrewmzhang**

### CODE REVIEW

**Comment by andrewmzhang** [comment]
Contains Code in multiple PS. I advise merging that one first

**Comment by andrewmzhang** [comment]
PR is ready for review

File: python/frontend/cirrus/cirrus/GridSearch.py [hunk 1]
```diff
@@ -52,6 +75,13 @@ def set_task_parameters(self, task, param_base=None, hyper_vars=[], hyper_params
         base_port = 1337
         index = 0
         num_machines = len(machines)
+
+        lc = boto3.client('lambda')
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/GridSearch.py]: Use better name

File: python/frontend/cirrus/cirrus/HyperParamDashDemo.ipynb [hunk 2]
```diff
@@ -0,0 +1,141 @@
+{
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/HyperParamDashDemo.ipynb]: I think this file shouldn't have been changed. Also the content doesn't look good

```
"name 'app' is not defined",
```

File: python/frontend/cirrus/cirrus/core.py [hunk 3]
```diff
@@ -118,17 +127,17 @@ def get_updates_per_second(self, fetch=True):
             t = time.time() - self.start_time
             ups = messenger.get_num_updates(self.ps_ip_public, self.ps_ip_port)
             self.time_ups_lst.append((t, ups))
+
+            cost_per_second = self.cost_model.get_cost(t)
+            if ups is not None and not (ups == 0):
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/core.py]: Why not ```ups != 0```?

File: python/frontend/cirrus/cirrus/core.py [hunk 4]
```diff
@@ -207,7 +216,7 @@ def launch_error_task(self, command_dict=None):
 
     def launch_ps(self, command_dict=None):
         cmd = 'nohup ./parameter_server --config config_%d.txt --nworkers %d --rank 1 --ps_port %d &> ps_out_%d & ' % (
-            self.ps_ip_port, self.n_workers * 2, self.ps_ip_port, self.ps_ip_port)
+            self.ps_ip_port, self.n_workers * 100, self.ps_ip_port, self.ps_ip_port)
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/core.py]: Why is this ```* 100```?

File: python/frontend/cirrus/cirrus/core.py [hunk 5]
```diff
@@ -224,3 +233,93 @@ def copy_config(self, command_dict=None):
     @abstractmethod
     def define_config(self, fetch=False):
         pass
+
+
+    def relaunch_lambdas(self):
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/core.py]: This function deserves a comment

File: python/frontend/cirrus/cirrus/core.py [hunk 6]
```diff
@@ -224,3 +233,93 @@ def copy_config(self, command_dict=None):
...
+        if rtl:
+            return self.fetch_metric(self.REAL_TIME_LOSS_VS_TIME)
+        else:
+            return self.fetch_metric(self.LOSS_VS_TIME)
+
+
+    def fetch_metric(self, key):
+        return self.metrics[key]
+
+    def maintain_error(self):
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/core.py]: This one also deserves comment.

File: python/frontend/cirrus/cirrus/messenger.py [hunk 7]
```diff
@@ -27,8 +27,8 @@ def get_last_time_error(ip="127.0.0.1", port=1338):
         clientsocket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
         clientsocket.sendto(GET_LAST_TIME_ERROR, (ip, port))
         clientsocket.settimeout(10)
-        s = clientsocket.recv(192)
-        return struct.unpack("ddd", s)
+        s = clientsocket.recv(256)
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/messenger.py]: This 256 probably should have a comment

File: python/frontend/cirrus/examples/HyperParamDashDemo.ipynb [hunk 8]
```diff
@@ -26,12 +26,47 @@
    "cell_type": "code",
    "execution_count": null,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
```
Comment (jcarreira) [inline python/frontend/cirrus/examples/HyperParamDashDemo.ipynb]: This file has unnecessary changes.

File: src/ErrorSparseTask.cpp [hunk 9]
```diff
@@ -208,16 +205,15 @@ void ErrorSparseTask::run(const Configuration& config) {
                   << " time(us): " << get_time_us()
                   << " time from start (sec): " << last_time << std::endl;
       } else if (config.get_model_type() == Configuration::COLLABORATIVE_FILTERING) {
-        last_error = std::sqrt(total_loss / total_num_features);
+        last_error = std::sqrt(total_loss / total_num_samples);
```
Comment (jcarreira) [inline src/ErrorSparseTask.cpp]: I think we discussed this before. The collaborative filtering loss is the RMSE of each value in the matrix, so I think this should be ```total_num_features```.

File: src/PSSparseServerTask.cpp [hunk 10]
```diff
@@ -576,7 +581,7 @@ void PSSparseServerTask::loop(int poll_id) {
   struct sockaddr_in cli_addr;
   socklen_t clilen = sizeof(cli_addr);
 
-  buffer.resize(10 * 1024 * 1024); // reserve 10MB upfront
+  // buffer.resize(10 * 1024 * 1024); // reserve 10MB upfront
```
Comment (jcarreira) [inline src/PSSparseServerTask.cpp]: Isn't ```buffer``` used?

File: src/SparseLRModel.cpp [hunk 11]
```diff
@@ -69,17 +69,16 @@ uint64_t SparseLRModel::getSerializedSize() const {
 }
 
 /** FORMAT
-  * number of weights (int)
-  * list of weights: weight1 (FEATURE_TYPE) | weight2 (FEATURE_TYPE) | ..
-  */
+ *   * number of weights (int)
```
Comment (jcarreira) [inline src/SparseLRModel.cpp]: Badly formatted.

File: python/frontend/cirrus/cirrus/GridSearch.py (hunk 1 above)
Comment (andrewmzhang) [reply python/frontend/cirrus/cirrus/GridSearch.py]: Apparently that's an unused variable. Removed and fixed.

File: python/frontend/cirrus/cirrus/HyperParamDashDemo.ipynb (hunk 2 above)
Comment (andrewmzhang) [reply python/frontend/cirrus/cirrus/HyperParamDashDemo.ipynb]: This file has been removed (this file DNE in master).

File: python/frontend/cirrus/cirrus/core.py (hunk 3 above)
Comment (andrewmzhang) [reply python/frontend/cirrus/cirrus/core.py]: ups is None when messenger.get_num_update errors. However the second clause in the condition is probably better as ups != 0. Fixed

File: src/PSSparseServerTask.cpp (hunk 10 above)
Comment (andrewmzhang) [reply src/PSSparseServerTask.cpp]: I dont think so. I can't see where its used. I'll double check this more thoroughly and report back

File: src/SparseLRModel.cpp (hunk 11 above)
Comment (andrewmzhang) [reply src/SparseLRModel.cpp]: Fixed the formatting

File: python/frontend/cirrus/examples/HyperParamDashDemo.ipynb (hunk 8 above)
Comment (andrewmzhang) [reply python/frontend/cirrus/examples/HyperParamDashDemo.ipynb]: Reverted

File: python/frontend/cirrus/cirrus/core.py (hunk 4 above)
Comment (andrewmzhang) [reply python/frontend/cirrus/cirrus/core.py]: Changing it back to 2. There is no reason that it should be 100.

File: python/frontend/cirrus/cirrus/messenger.py (hunk 7 above)
Comment (andrewmzhang) [reply python/frontend/cirrus/cirrus/messenger.py]: Num of bits of 4 doubles. Comment added.

File: src/ErrorSparseTask.cpp (hunk 9 above)
Comment (andrewmzhang) [reply src/ErrorSparseTask.cpp]: Reverted

File: python/frontend/cirrus/cirrus/core.py (hunk 6 above)
Comment (andrewmzhang) [reply python/frontend/cirrus/cirrus/core.py]: Comments made

File: python/frontend/cirrus/cirrus/core.py (hunk 5 above)
Comment (andrewmzhang) [reply python/frontend/cirrus/cirrus/core.py]: Comment made
